relayer/src/chain/cosmos/batch.rs

```diff
@@ -36,20 +36,15 @@ pub async fn send_batched_messages_and_wait_commit(
     )
     .await?;
 
-    wait_for_block_commits(
+    let events = wait_for_block_commits(
```

Comment:
The original intent for `send_batched_messages_and_wait_commit()` was:
- send `N` messages to the chain
- get back exactly `N` `IbcEvent`-s such that result `i` is for message `i`

Not sure this still holds. For non-IBC messages we could still get `IbcEvent::Empty`

Reply:
I don't understand the rationale to expect the number of `IbcEvent`s to be exactly the same as the number of messages sent. In theory one message can result in any number of chain events, which can be translated into any number of `IbcEvent`s.

I think a more practical approach is to use `find` through the result list of `IbcEvent`s to find the event of interest, instead of looking at a specific position.

Also the change does not affect the semantics of existing transactions sent. So if the existing code produce messages that result in exact number of `IbcEvent`s, then they would still get that. It would only affect future changes, such as when the code is updated to send non-IBC messages mixed with IBC messages.

Reply:
> I don't understand the rationale to expect the number of `IbcEvent`s to be exactly the same as the number of messages sent. In theory one message can result in any number of chain events, which can be translated into any number of `IbcEvent`s.

It's always been the case in practice that an IBC message results in one and only one IBC event. While the events have not been documented in the IBC spec yet, some documentation exists [here](https://github.com/cosmos/ibc-go/blob/main/docs/ibc/events.md)

> 
> I think a more practical approach is to use `find` through the result list of `IbcEvent`s to find the event of interest, instead of looking at a specific position.

You still need to figure out the original message that caused a specific `IbcEvent`.  But sure there are other ways of coding this. Still trying to understand the usecase that breaks the current design.

> 
> Also the change does not affect the semantics of existing transactions sent. So if the existing code produce messages that result in exact number of `IbcEvent`s, then they would still get that.

Yes I was asking in general not necessarily pointing to your changes in this PR. 

> It would only affect future changes, such as when the code is updated to send non-IBC messages mixed with IBC messages.

Yeah, i still want to understand this. I think hermes should only deal with IBC messages and never send messages that have nothing to do with IBC. Probably we should start with what "IBC message" means? And what is a non-IBC message?

Reply:
> The original intent for `send_batched_messages_and_wait_commit()` was:
> 
> * send `N` messages to the chain
> * get back exactly `N` `IbcEvent`-s such that result `i` is for message `i`
> 
> Not sure this still holds. For non-IBC messages we could get `IbcEvent::Empty`

i checked master and it is broken. When a Tx has N messages, if the Tx fails then all messages should be marked as failed and get back N error events. I have a quick fix in branch `anca/err_for_all_msgs_in_tx`. Also added and corrected some comments. 
I haven't checked/ fixed the packet worker (async sender), at first look it doesn't handle error events here: (https://github.com/informalsystems/ibc-rs/blob/194a1714cf9dc807dd7b4e33726319293848dc58/relayer/src/link/relay_path.rs#L609-L615). 
For the sync sender cases the code looks better.